### PR TITLE
[Analyzer] Dont use static:: in final class #164

### PR DESCRIPTION
**Pull request by veloxy**
Hey!

Type: new feature

Link to issue: https://github.com/ovr/phpsa/issues/164

This pull request affects the following components: **(please check boxes)**

* [ ] Core
* [x] Analyzer
* [ ] Compiler
* [ ] Control Flow Graph
* [ ] Documentation

**In raising this pull request, I confirm the following (please check boxes):**

- [x] I have read and understood the [Contributing Guidelines](/.github/CONTRIBUTING.md)?
- [x] I have checked that another pull request for this purpose does not exist.
- [x] I wrote some tests for this PR.

Small description of change: Checks wether static calls are used in final classes.

Thanks

### CODE REVIEW

File: src/Analyzer/Pass/Expression/FinalStaticUsage.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+{
+    /**
+     * @param Expr\StaticCall $expr
+     * @param Context $context
+     * @return bool
+     */
+    public function pass(Expr\StaticCall $expr, Context $context)
+    {
+        $classObject = $context->scope->getPointer()->getObject();
+
```
Comment (ovr) [inline src/Analyzer/Pass/Expression/FinalStaticUsage.php]: Remove space link because `$classObject` variable related to if

Comment (veloxy) [inline src/Analyzer/Pass/Expression/FinalStaticUsage.php]: Done!

**Comment by veloxy** [comment]
I'm not sure how to add a test without static usage that doesn't trigger the warning, and I would like to test the `isFinal` I added, any suggestions for that?

Also, non-related - Github is showing all my force pushes - can I prevent that somehow?

**Comment by ddmler** [comment]
If you mean force pushes here in PR then just reload and they are gone. if you mean the messages in the issue then you should not write the issue number in your commit message (this generates the message)

For testing the analyzer when not using static:: you could for example use self:: or just leave that test out.
For testing isFinal: you could create a new Node\Stmt\Class_ with final attribute. I have an idea how we could make this easier I will look into it tomorrow. For now you can leave that test out i think.

**Comment by veloxy** [comment]
@ddmler I wanted to add just a final class but didn't know what to put in the json - as far as I could see it always assumed `true`. I guess I'll leave it out for know, let me know if there's anything else I need to do/change!

**Comment by ddmler** [comment]
@veloxy you would just add the empty final class below the other and change nothing in json. (because there is no new error) the test runner checks that there are no other notices than in json with same type.

File: src/Analyzer/Pass/Expression/FinalStaticUsage.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+class FinalStaticUsage implements AnalyzerPassInterface
+{
+    /**
+     * @param Expr\StaticCall $expr
+     * @param Context $context
+     * @return bool
+     */
+    public function pass(Expr\StaticCall $expr, Context $context)
+    {
+        $classObject = $context->scope->getPointer()->getObject();
```
Comment (ddmler) [inline src/Analyzer/Pass/Expression/FinalStaticUsage.php]: `$context->scopePointer->getObject();`
The context has already a scopePointer. I'll have to check what scope is for or if we can get rid of it

Comment (veloxy) [inline src/Analyzer/Pass/Expression/FinalStaticUsage.php]: I believe the scopePointer didn't contain the class so I couldn't check if the class was final, eventually I found it in the scope.

**Comment by ddmler** [comment]
Ok Thanks @veloxy i will look into the thing with scopePointer.

**Pull request merged**